<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-3.0.1"><span class="toc-item-num">3.0.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-4.0.1"><span class="toc-item-num">4.0.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод:</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

# Цель исследования
1. Разработать такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию и обосновать корректность его работы.


# Ход исследования
Нам предоставленны данные клиентов страховой компании «Хоть потоп». Данные находятся в датафрейме, который мы получим из файла '/datasets/insurance.csv'. О качестве данных нам ничего не известно, поэтому перед тем, как приступать к целям исследования, понадобится обзор данных. Мы проверим данные на ошибки. Затем, на этапе предобработки мы будем искать возможность исправить все ошибки в данных, которые не приведут к искажению конечного результата. Далее приступим к целям исследования.<br>


Таким образом, исследование пройдёт в семь этапов:<br>

* Обзор данных <br>
* Предобработка данных <br>
* Ответ на вопрос об умножении матриц <br>
* Написание алгоритма преобразования<br>
* Проверка алгоритма <br>
* Написание промежуточных выводов <br>
* Написание общего вывода

## Загрузка данных

In [1]:
#Импортируем необходимые библиотеки
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.metrics import r2_score

In [2]:
#Загружаем и изучаем датафрейм
try:
    data = pd.read_csv('/datasets/insurance.csv')
except:    
    pass

In [24]:
#Ознакамливаемся с датареймом
data.head(10)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
5    1     43.0   41000.0            2                  1
6    1     39.0   39700.0            2                  0
7    1     25.0   38600.0            4                  0
8    1     36.0   49700.0            1                  0
9    1     32.0   51700.0            1                  0

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [26]:
#Проверяем количество пропусков в %
pd.DataFrame(round(data.isna().mean()*100)).style.background_gradient('coolwarm')

In [27]:
#Проверяем количество явных дубликатов
print('Количество дубликатов:', data.duplicated().sum())

Количество дубликатов: 153


In [28]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [29]:
#проверяем уникальные значения
for col in data:
    print(data[col].unique())

[1 0]
[41. 46. 29. 21. 28. 43. 39. 25. 36. 32. 38. 23. 40. 34. 26. 42. 27. 33.
 47. 30. 19. 31. 22. 20. 24. 18. 37. 48. 45. 44. 52. 49. 35. 56. 65. 55.
 57. 54. 50. 53. 51. 58. 59. 60. 61. 62.]
[49600. 38000. 21000. 41700. 26100. 41000. 39700. 38600. 49700. 51700.
 36600. 29300. 39500. 55000. 43700. 23300. 48900. 33200. 36900. 43500.
 36100. 26600. 48700. 40400. 38400. 34600. 34800. 36800. 42200. 46300.
 30300. 51000. 28100. 64800. 30400. 45300. 38300. 49500. 19400. 40200.
 31700. 69200. 33100. 31600. 34500. 38700. 39600. 42400. 34900. 30500.
 24200. 49900. 14300. 47000. 44800. 43800. 42700. 35400. 57200. 29600.
 37400. 48100. 33700. 61800. 39400. 15600. 52600. 37600. 52500. 32700.
 51600. 60900. 41800. 47400. 26500. 45900. 35700. 34300. 26700. 25700.
 33300. 31100. 31500. 42100. 37300. 42500. 27300. 46800. 33500. 44300.
 41600. 53900. 40100. 44600. 45000. 32000. 38200. 33000. 38500. 51800.
 33800. 46400. 43200. 31800. 50200. 35100. 30700. 45800. 49300. 42800.
 33600. 50300. 34000. 364

In [30]:
#Находим повторяющиеся строки во всем датафрейме
data[data.duplicated()]

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
281     1     39.0   48100.0            1                  0
488     1     24.0   32900.0            1                  0
513     0     31.0   37400.0            2                  0
718     1     22.0   32600.0            1                  0
785     0     20.0   35800.0            0                  0
...   ...      ...       ...          ...                ...
4793    1     24.0   37800.0            0                  0
4902    1     35.0   38700.0            1                  0
4935    1     19.0   32700.0            0                  0
4945    1     21.0   45800.0            0                  0
4965    0     22.0   40100.0            1                  0

[153 rows x 5 columns]

In [10]:
#Удаляем и проверяем датафрейм на дубликаты 
data = data.drop_duplicates(keep=False)
data.duplicated().sum()

0

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4700 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                4700 non-null   int64  
 1   Возраст            4700 non-null   float64
 2   Зарплата           4700 non-null   float64
 3   Члены семьи        4700 non-null   int64  
 4   Страховые выплаты  4700 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 220.3 KB


#### Вывод:
Мы загрузили и ознакомились с представленным датафреймом, всего в датафрейме 5 столбцов и 5 000 записей, пропусков не выялено. <br>
* Признаки: пол, возраст и зарплата застрахованного, количество членов его семьи;
* Целевой признак: количество страховых выплат клиенту за последние 5 лет.

Обнаружено 153 дубликата, которые мы удалили, теперь в датафрейме всего 4 700 записи. Тип данных корректен.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** *b. Не изменится.*

**Обоснование:**

Для обоснования ответа получим предсказания для домноженных признаков. Пусть наши признаки будут как в исходном датасете размерностью 4700 на 4, а матрица для домножения 4 на 4. Умноженная матрица будет иметь вид X * P. Тогда фомула предсказаний примет следующий вид:
$$ a' = XPw' $$

$$ a' = X'w' = XPw'$$

$$ w' = (X'^T X')^{-1} X'^T y = ((XP)^T XP)^{-1})(XP)^T y $$

Подставим в формулу предсказаний формулу обучения для матрицы $ X P $:
$$ a' = XP((XP)^T XP)^{-1})(XP)^T y $$

Запишем свойства матриц и воспользуемся ими:
$$ (AB)^T = B^T A^T $$
$$ (AB)^{-1} = B^{-1} A^{-1} $$
$$ AA^{-1} = A^{-1} A = E $$
$$ AE = EA = A $$

Раскроем скобки с транспонированием матриц, получится следующее уравнение:
$$ a' = XP(P^T X^T XP)^{-1} P^T X^T y $$

Внутри скобок, от множителей которых берется обратная матрица остались следующие матрицы: $ P^T $ размером 4х4, $ X^T $ размером 4х4700,$ X $ размером 4700x4,$ P $ размером 4x4. Т.к. обратная матрица может браться только от квадратной матрицы, то для раскрытия скобок данного произведения нам надо сгруппировать его множители так, чтобы они были квадратными матрицами.

Выделим 2 множителя произведения $ P^T X^T X $ и $ P $ и раскроем скобки:

$$ a' = XPP^{-1}(P^T X^T X)^{-1} P^T X^T y $$
 
Выделим внутри оставшихся скобок множители $ P^T $ и $ X^T X $ и раскроем скобки: 
$$ a' = XPP^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y $$

Произведения $ PP^{-1} $ и $ (P^T)^{-1} P^T $ дают единичные матрицы, которые можно исключить из уравнения. С учетом этого формула принимает вид:
$$ a' = X (X^T X)^{-1} X^T y $$

Можно заметить, что произведение равно фомуле предсказания, поэтому получаем:

$$ a' = Xw $$

Мы получили аналогичную формулу предсказаний, что означает, что качество линейной регрессии не изменится, так как формула исходного предсказания равна формуле предказания, когда признаки домножили на обратимую матрицу, чтд. 

#### Вывод:
Мы ответили на вопрос и обосновали решение "Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?" - качество линейной регрессии не изменится, так как формула исходного предсказания равна формуле предказания, когда признаки умножили на обратимую матрицу.

## Алгоритм преобразования

**Алгоритм**

В соответствии с алгоритом преобразования, умножим матрицу признаков на обратимую матрицу. <br>

* Умножение матрицы на матрицу возможно, если ширина первой матрицы А$(mxn)$ равна высоте второй матрицы P$(nxr)$. Тогда размер произведения этих матриц будет $mxr$, размерность $n$ «cхлопывается». <br>

* При умножении признаков выборки размерностью $(4700, 4)$ на обратимую произвольную матрицу размерностью (4, 4), в результате получаем новый набор данных скалярных произведений.

In [12]:
#Объявим две переменные 
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']

In [13]:
#Создадим алгоритм для рандомной матрицы
class encryptedModel:
    def fit(self, train_features, train_target):
        self.R = np.linalg.inv(np.random.rand(4,4))
        self.M = train_features @ self.R
        X = np.concatenate((np.ones((self.M.shape[0], 1)), self.M), axis=1)
        y = train_target
        w = ((np.linalg.inv(X.T.dot(X))) @ X.T) @ y
        self.w = w[1:]
        self.w0 = w[0]
    def predict(self, test_features):
        return (test_features @ self.R) @ self.w + self.w0
    
model = encryptedModel()
model.fit(features, target)
predictions = model.predict(features)
print('Значение R2:', r2_score(target, predictions))   

Значение R2: 0.43553667106062344


**Обоснование**

Так как мы доказали, что при умножении на обратимую матрицу качество линейной регрессии не ухудшится, то создали модель линейной регрессии так, чтобы она создавала рандомную обратимую матрицу, а так же высчитывала весы и предсказания.

#### Вывод:
* Мы предложили алгоритм преобразования данных для решения задачи;
* Обосновали, почему качество линейной регрессии не поменяется.

## Проверка алгоритма

In [14]:
#Создадим класс линейной регрессии и проверим алгоритм на исходных данных (признаках)
class LinearRegression:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = np.linalg.inv(X.T @ X) @ X.T @ y 
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0
    
model = LinearRegression()
model.fit(features, target)
predictions = model.predict(features)
print('Значение R2 на исходных данных:', r2_score(target, predictions))

Значение R2 на исходных данных: 0.4355366710606289


In [15]:
#Проверим алгоритм на данных (признаках) умноженных на обратимую матрицу
model_2 = encryptedModel()
model_2.fit(features, target)
predictions_2 = model_2.predict(features)
print('Значение R2 данных умноженных на обратимую матрицу:', r2_score(target, predictions_2))

Значение R2 данных умноженных на обратимую матрицу: 0.43553667106012994


#### Вывод:
* Мы запрограммировали алгоритм, применив матричные операции;
* Проверили, что качество линейной регрессии из sklearn не отличается до и после преобразования и применили метрику R2:
    * Значения R2 на исходных данных и данных, умноженных на обратимую матрицу, одинаковы и составляют 0.43. 
    * Алгорим проверен и работает коректно, модели показали одинаковые результаты с пребразованием и без. 

## Общий вывод:
1. Мы ознакомились с представленным датафреймом и провели предобработку данных.
<br>
2. Мы ответили на вопрос и обосновали решение "Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?" - качество линейной регрессии не изменится, так как формула исходного предсказания равна формуле предказания, когда признаки умножили на обратимую матрицу.<br>
3. Мы предложили алгоритм преобразования данных для решения задачи.<br>
4. Обосновали, почему качество линейной регрессии не поменяется.
<br>
5. Мы запрограммировали алгоритм, применив матричные операции.<br>
6. Проверили, что качество линейной регрессии из sklearn не отличается до и после преобразования и применили метрику R2:
    * Значения R2 на исходных данных и данных, умноженных на обратимую матрицу, одинаковы и составляют 0.43. 
    * Алгорим проверен и работает коректно, модели показали одинаковые результаты с пребразованием и без. <br> 
   
*Рекомендации:*<br> 
Был разработан такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию, соответственно, данные клиентов страховой компании «Хоть потоп» будут надежно защищены, что положительно отразится на лояльности клиентов, повысит репутацию компании и привлечет еще больше клиентов. 